### PR TITLE
test: Split huge monolith mock header to speed up test compilation

### LINKED ISSUE #10917: Reduce resource consumption of test compilation by simplifying mock library inclusions
*Description*:
Many tests/test libraries over-include especially mock libraries which causes increased memory usage, disk usage, and compilation time. This is especially a problem on Windows with `cl.exe` but we have also observed significant resource usage that can be reduced by as much as half on Linux with `clang`.

Attached to this issue is a file showing the peak working set memory usage of `cl.exe` compiling various tests/test libraries (units in KB). We can see many of the tests cause the compiler to use multiple GB of memory, often for simple tests that do not actually require many of the includes they are being built with. We have been able to reduce this usage (see linked PR) by cleaning up these includes. Our primary focus has been enabling faster/less resource intensive compilation of tests in Windows CI but this endeavor seems like it would be relevant to compilation performance on other platforms. We can potentially reduce compilation time, disk usage of artifacts/parameter files, and memory usage of the compiler (by orders of magnitude in some cases, multiple GB reduction in others).

While we have had some success empirically assessing the memory usage of compiling tests and finding the most expensive, it is not super feasible to go one-by-one and manually assess/fix the includes etc. to pare-down each test. We are hoping to potentially get some help reducing the overhead of compiling these tests from other contributors working toward similar changes as in the PR linked below as well as finding a programmatic solution that could potentially be automated (in CI or otherwise) that could identify over-inclusion of mocks and other large libraries that are not used. Other strategies to help solve this could be further breaking down larger libraries with many classes into their component parts so tests only include exactly what interfaces etc. they need.

*Relevant Links*:
- See https://github.com/envoyproxy/envoy/pull/10915 for a relevant PR that addresses this for the `kafka` network filter extension tests
- Gist with peak working set memory usage for tests that exceed 2GB: https://gist.github.com/sunjayBhatia/d2d4a0af1eab16777f94a4dc4959b342
- Powershell script used to collect data: https://github.com/greenhouse-org/envoy/blob/84fca545722432a48371fca1ced02253cddb86d0/get-cl-workingset-peak.ps1

### PR DESCRIPTION
**Pull request by foreseeable**
Commit Message: Split huge monolith mock header to speed up test compilation
Additional Description: `cluster_manager_test` only used a simple mock class `MockAdmin` from `test/mocks/server/mocks.h`, which is a huge mock library. After splitting, the overall build time for `cluster_manager_test` reduced from 143.481s to 82.443s in my build cluster.

Risk Level: low
Testing: existing tests
Docs Changes: N/A
Release Notes: no
Related Issues: #10917

### CODE REVIEW

**Comment by ahedberg** [comment]
@asraa FYI

File: test/mocks/server/admin.cc [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+using testing::_;
+using testing::Invoke;
+using testing::Return;
+using testing::ReturnPointee;
+using testing::ReturnRef;
+using testing::SaveArg;
```
Comment (ahedberg) [inline test/mocks/server/admin.cc:18]: We're not using any of these using declarations in this file, so they can be deleted. This is part of why the clang-tidy check is failing.

File: test/mocks/server/config_tracker.cc [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+using testing::_;
+using testing::Invoke;
+using testing::Return;
+using testing::ReturnPointee;
+using testing::ReturnRef;
+using testing::SaveArg;
```
Comment (ahedberg) [inline test/mocks/server/config_tracker.cc:18]: Same here - we're not using these using declarations, so they can be removed.

File: test/mocks/server/config_tracker.h [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+#include "test/mocks/secret/mocks.h"
+#include "test/mocks/stats/mocks.h"
+#include "test/mocks/thread_local/mocks.h"
+#include "test/mocks/tracing/mocks.h"
+#include "test/mocks/upstream/mocks.h"
+#include "test/test_common/test_time_system.h"
+
+#include "absl/strings/string_view.h"
+#include "gmock/gmock.h"
+#include "spdlog/spdlog.h"
```
Comment (ahedberg) [inline test/mocks/server/config_tracker.h:57]: Can we do a pass over the headers here and in admin.h to make sure we're only including what we need? We can go over this during a 1:1 if that would help.

**Comment by foreseeable** [comment]
The overall build time for cluster_manager_test has been further improved to 68.689s in my build cluster by removing superfluous includes in divided headers.

File: test/mocks/server/admin.h [hunk 4]
```diff
@@ -1,61 +1,14 @@
...
-#include "test/mocks/tracing/mocks.h"
-#include "test/mocks/upstream/mocks.h"
-#include "test/test_common/test_time_system.h"
 
 #include "absl/strings/string_view.h"
 #include "config_tracker.h"
 #include "gmock/gmock.h"
-#include "spdlog/spdlog.h"
+
+using testing::NiceMock;
```
Comment (ahedberg) [inline test/mocks/server/admin.h:11]: Using-declarations should only be used in .cc files. See https://google.github.io/styleguide/cppguide.html#Namespaces

File: test/mocks/server/config_tracker.cc [hunk 5]
```diff
@@ -2,14 +2,12 @@
...
 
-#include "envoy/admin/v3/server_info.pb.h"
-#include "envoy/config/core/v3/base.pb.h"
-
-#include "common/singleton/manager_impl.h"
-
 #include "gmock/gmock.h"
 #include "gtest/gtest.h"
 
+using testing::_;
```
Comment (ahedberg) [inline test/mocks/server/config_tracker.cc:8]: These should be moved to inside the innermost namespace, and they should be fully-qualified. See https://abseil.io/tips/119

File: test/mocks/server/admin.h (hunk 4 above)
Comment (foreseeable) [reply test/mocks/server/admin.h]: done

File: test/mocks/server/config_tracker.cc (hunk 5 above)
Comment (foreseeable) [reply test/mocks/server/config_tracker.cc]: fixed

**Comment by sunjayBhatia** [comment]
Just a nit on the PR description, the comment `Partly Fixes #10917` will close the original issue (if you hover over `Fixes`), I would guess there is additional work to be done for that issue, should we adjust that part of the PR description?

**Comment by foreseeable** [comment]
@sunjayBhatia Thanks for your advice. Now the PR description has been adjusted to "Related Issues"

**Pull request merged**